torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1472,44 +1472,68 @@ void ProcessGroupNCCL::shutdown() {
 
 // NOLINTNEXTLINE(bugprone-exception-escape)
 ProcessGroupNCCL::~ProcessGroupNCCL() {
-  VLOG(2) << logPrefix() << "ProcessGroupNCCL destructor entered.";
-
-  if (!terminateProcessGroup_.load()) {
-    if (rank_ % localDeviceCount_ == 0) {
-      TORCH_WARN_ONCE(
-          "WARNING: process group has NOT been destroyed before we destruct ProcessGroupNCCL. ",
-          "On normal program exit, the application should call destroy_process_group to ",
-          "ensure that any pending NCCL operations have finished in this process. "
-          "In rare cases this process can exit before this point and block the progress of "
-          "another member of the process group. This constraint has always been present, "
-          "but this warning has only been added since PyTorch 2.4");
-    }
-    // If user haven't explicitly destroy/shutdown process group, destructor
-    // needs to do so
-    // Note: we have rewritten `shutdown` to represent the destroy behavior.
-    // Here we route to `abort()` explicitly to maintain the old behavior, until
-    // we fix everything.
+  LOG(INFO) << logPrefix() << "ProcessGroupNCCL destructor entered.";
+  bool cudaContextAlive;
+
+  if (terminateProcessGroup_.load())
+    // `shutdown()` or `abort` already called. Skip the favor of disposing
+    // communicators.
+    goto join_threads;
+
+  // If user haven't explicitly destroy/shutdown process group, destructor
+  // needs to do so
+  // First print warning on first rank of each node
+  if (rank_ % localDeviceCount_ == 0) {
+    TORCH_WARN_ONCE(
+        "WARNING: process group has NOT been destroyed before we destruct ProcessGroupNCCL. ",
```

Comment:
nit: maybe we should shorten this message too. I think the second half is not needed.  We could also poitn users to a docpage that has more detail if we need the detail.  I'd like it better if it were more like 

"WARNING: destroy_process_group() was not called before shutdown, which can leak resources.  For more info, see ..."

Reply:
Opened https://github.com/pytorch/pytorch/pull/142297 for messaging improvement.